<a href="https://colab.research.google.com/github/Prebor97/stageBcodes/blob/master/Prebors_modelling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving finalDatasetWithRain.csv to finalDatasetWithRain.csv


In [ ]:
! pip install plotly

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
import plotly.express as px

In [ ]:
import pandas as pd
import numpy as np
import io
df = pd.read_csv(io.BytesIO(uploaded['finalDatasetWithRain.csv']))

In [ ]:
def Evaluation_metrics(predicted, actual):
  MAE = np.mean(abs(predicted - actual))
  MSE = np.square(np.subtract(predicted, actual)).mean()
  RMSE = np.sqrt(np.mean(np.square(predicted - actual)))
  MAPE = np.mean(np.abs(np.subtract(predicted, actual)))
  print('The Mean Absolute Error is ', round(MAE,2))
  print('The Mean Squared Error is ', round(MSE,2))
  print('The Root Mean Squared Error is ', round(RMSE,2))
  print('The Mean Absolute Percentage Error is ', round(MAPE,2))

I am filtering the third location 

In [ ]:
df1 = df.iloc[51806:77709]

In [ ]:
df1.reset_index(inplace = True, drop = True)
df1.head()

lat   lon        time   omega_x   omega_y     omega     pr_wtr  rhum_x  \
0  30.0  75.0  1948-01-01 -0.103003 -0.161001 -0.030003  12.429993   19.75   
1  30.0  75.0  1948-01-02 -0.014002 -0.043001 -0.012003  16.989990   11.75   
2  30.0  75.0  1948-01-03 -0.047003 -0.024002  0.103998  18.470001   19.50   
3  30.0  75.0  1948-01-04 -0.005003 -0.073002 -0.084002  17.019989   15.75   
4  30.0  75.0  1948-01-05  0.133999  0.137999  0.178997  13.940002   19.00   

   rhum_y       rhum  ...      tmp_x   tmp_y        tmp     uwnd_x    uwnd_y  \
0   38.75  34.500000  ...  297.59998  285.47  296.20000  14.300003  2.390015   
1   46.00  61.000015  ...  297.59998  287.55  296.12000  15.420013 -1.259994   
2   40.25  73.500015  ...  297.59998  287.45  296.09998  13.070007 -0.189987   
3   25.25  62.500015  ...  297.59998  287.05  296.07000   7.790008 -0.259995   
4   12.25  57.250015  ...  297.59998  286.57  295.99000  19.050003  3.150009   

       uwnd     vwnd_x    vwnd_y      vwnd      rain  
0 -3.789993   7.170013  6.250015  1.250015   0.00000  
1 -4.139984  -0.189987  2.570007 -0.419998  46.66959  
2 -4.209992   2.120010  1.450012 -3.109985   0.00000  
3 -1.379990  -2.389984  2.370010 -0.659988   0.00000  
4 -3.259994 -16.139984 -4.839996 -4.559998   0.00000  

[5 rows x 21 columns]

In [ ]:
cor_df = df1.drop(columns = 'time')

In [ ]:
cor_df.corr()

lat  lon   omega_x   omega_y     omega    pr_wtr    rhum_x    rhum_y  \
lat      NaN  NaN       NaN       NaN       NaN       NaN       NaN       NaN   
lon      NaN  NaN       NaN       NaN       NaN       NaN       NaN       NaN   
omega_x  NaN  NaN  1.000000  0.863237  0.631888 -0.176563 -0.274799 -0.357339   
omega_y  NaN  NaN  0.863237  1.000000  0.749932  0.018964 -0.202758 -0.315366   
omega    NaN  NaN  0.631888  0.749932  1.000000 -0.290581 -0.128694 -0.358164   
pr_wtr   NaN  NaN -0.176563  0.018964 -0.290581  1.000000  0.117803  0.482833   
rhum_x   NaN  NaN -0.274799 -0.202758 -0.128694  0.117803  1.000000  0.494687   
rhum_y   NaN  NaN -0.357339 -0.315366 -0.358164  0.482833  0.494687  1.000000   
rhum     NaN  NaN -0.248803 -0.076290 -0.217727  0.723536  0.106789  0.273418   
slp      NaN  NaN  0.108792  0.021456  0.347598 -0.768066  0.113635 -0.316519   
tmp_x    NaN  NaN  0.051662  0.034606 -0.010568  0.003696 -0.032935 -0.044126   
tmp_y    NaN  NaN -0.021705  0.037248 -0.253736  0.709890 -0.118593  0.350184   
tmp      NaN  NaN  0.036800  0.191021 -0.043212  0.637207 -0.204841 -0.012373   
uwnd_x   NaN  NaN -0.012728 -0.213734  0.037093 -0.748486  0.034582 -0.213907   
uwnd_y   NaN  NaN  0.037769 -0.192428  0.002886 -0.699865  0.091913 -0.106991   
uwnd     NaN  NaN  0.234067  0.152107 -0.102317 -0.067077 -0.198795 -0.113265   
vwnd_x   NaN  NaN -0.708233 -0.560227 -0.329611  0.097893  0.171662  0.216085   
vwnd_y   NaN  NaN -0.766607 -0.797133 -0.459590 -0.104404  0.241165  0.256770   
vwnd     NaN  NaN -0.674882 -0.764811 -0.767462  0.288286  0.142932  0.429780   
rain     NaN  NaN -0.130614 -0.075272 -0.105365  0.258819  0.093132  0.146258   

             rhum       slp     tmp_x     tmp_y       tmp    uwnd_x    uwnd_y  \
lat           NaN       NaN       NaN       NaN       NaN       NaN       NaN   
lon           NaN       NaN       NaN       NaN       NaN       NaN       NaN   
omega_x -0.248803  0.108792  0.051662 -0.021705  0.036800 -0.012728  0.037769   
omega_y -0.076290  0.021456  0.034606  0.037248  0.191021 -0.213734 -0.192428   
omega   -0.217727  0.347598 -0.010568 -0.253736 -0.043212  0.037093  0.002886   
pr_wtr   0.723536 -0.768066  0.003696  0.709890  0.637207 -0.748486 -0.699865   
rhum_x   0.106789  0.113635 -0.032935 -0.118593 -0.204841  0.034582  0.091913   
rhum_y   0.273418 -0.316519 -0.044126  0.350184 -0.012373 -0.213907 -0.106991   
rhum     1.000000 -0.290428 -0.024584  0.220392  0.476342 -0.469037 -0.522365   
slp     -0.290428  1.000000  0.007890 -0.903137 -0.453904  0.647434  0.524864   
tmp_x   -0.024584  0.007890  1.000000  0.025606  0.081019 -0.007404 -0.024630   
tmp_y    0.220392 -0.903137  0.025606  1.000000  0.394563 -0.653045 -0.491136   
tmp      0.476342 -0.453904  0.081019  0.394563  1.000000 -0.548825 -0.636141   
uwnd_x  -0.469037  0.647434 -0.007404 -0.653045 -0.548825  1.000000  0.881638   
uwnd_y  -0.522365  0.524864 -0.024630 -0.491136 -0.636141  0.881638  1.000000   
uwnd    -0.165615 -0.164978 -0.002733  0.143734 -0.081981  0.088407  0.303142   
vwnd_x   0.146113 -0.010139 -0.001482 -0.012335  0.046966  0.038574 -0.088086   
vwnd_y   0.031428  0.146836  0.009636 -0.160708 -0.205517  0.259330  0.136362   
vwnd     0.230960 -0.316005  0.017627  0.269819  0.046749 -0.079272 -0.068615   
rain     0.262295 -0.150566 -0.012169  0.095727  0.127223 -0.147309 -0.162890   

             uwnd    vwnd_x    vwnd_y      vwnd      rain  
lat           NaN       NaN       NaN       NaN       NaN  
lon           NaN       NaN       NaN       NaN       NaN  
omega_x  0.234067 -0.708233 -0.766607 -0.674882 -0.130614  
omega_y  0.152107 -0.560227 -0.797133 -0.764811 -0.075272  
omega   -0.102317 -0.329611 -0.459590 -0.767462 -0.105365  
pr_wtr  -0.067077  0.097893 -0.104404  0.288286  0.258819  
rhum_x  -0.198795  0.171662  0.241165  0.142932  0.093132  
rhum_y  -0.113265  0.216085  0.256770  0.429780  0.146258  
rhum    -0.165615  0.146113  0.031428  0.230960  0.262295  
slp  

#### From the correlation we have above, we see that 6 features have good correlation with the target and they include : rhum(0.7235), slp (0.768066),tmp_y(0.709890),tmp(0.637207),uwnd_x(0.748486) and uwnd_y(0.699865). I am going to make use of these 6 features in my multivariate forecast

In [ ]:
df1['time']= pd.to_datetime(df1['time'])
df1 = df1.rename(columns = {'time': 'ds', 'pr_wtr' : 'y'})

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
df1.head()

lat   lon         ds   omega_x   omega_y     omega          y  rhum_x  \
0  30.0  75.0 1948-01-01 -0.103003 -0.161001 -0.030003  12.429993   19.75   
1  30.0  75.0 1948-01-02 -0.014002 -0.043001 -0.012003  16.989990   11.75   
2  30.0  75.0 1948-01-03 -0.047003 -0.024002  0.103998  18.470001   19.50   
3  30.0  75.0 1948-01-04 -0.005003 -0.073002 -0.084002  17.019989   15.75   
4  30.0  75.0 1948-01-05  0.133999  0.137999  0.178997  13.940002   19.00   

   rhum_y       rhum  ...      tmp_x   tmp_y        tmp     uwnd_x    uwnd_y  \
0   38.75  34.500000  ...  297.59998  285.47  296.20000  14.300003  2.390015   
1   46.00  61.000015  ...  297.59998  287.55  296.12000  15.420013 -1.259994   
2   40.25  73.500015  ...  297.59998  287.45  296.09998  13.070007 -0.189987   
3   25.25  62.500015  ...  297.59998  287.05  296.07000   7.790008 -0.259995   
4   12.25  57.250015  ...  297.59998  286.57  295.99000  19.050003  3.150009   

       uwnd     vwnd_x    vwnd_y      vwnd      rain  
0 -3.789993   7.170013  6.250015  1.250015   0.00000  
1 -4.139984  -0.189987  2.570007 -0.419998  46.66959  
2 -4.209992   2.120010  1.450012 -3.109985   0.00000  
3 -1.379990  -2.389984  2.370010 -0.659988   0.00000  
4 -3.259994 -16.139984 -4.839996 -4.559998   0.00000  

[5 rows x 21 columns]

In [ ]:
df1.columns

Index(['lat', 'lon', 'ds', 'omega_x', 'omega_y', 'omega', 'y', 'rhum_x',
       'rhum_y', 'rhum', 'slp', 'tmp_x', 'tmp_y', 'tmp', 'uwnd_x', 'uwnd_y',
       'uwnd', 'vwnd_x', 'vwnd_y', 'vwnd', 'rain'],
      dtype='object')

In [ ]:
x = df1.drop(columns = ['lat','lon', 'omega_x','omega_y','omega','rhum_x', 'rhum_y','tmp_x','uwnd', 'vwnd_x','vwnd_y','vwnd','rain'])

In [ ]:
x.columns

Index(['ds', 'y', 'rhum', 'slp', 'tmp_y', 'tmp', 'uwnd_x', 'uwnd_y'], dtype='object')

In [ ]:
x = x.rename(columns = {'rhum': 'add1', 'slp' : 'add2', 'tmp_y': 'add3', 'tmp' : 'add4', 'uwnd_x' : 'add5', 'uwnd_y' : 'add6'})

In [ ]:
x.head()

ds          y       add1      add2    add3       add4       add5  \
0 1948-01-01  12.429993  34.500000  102027.0  285.47  296.20000  14.300003   
1 1948-01-02  16.989990  61.000015  102087.0  287.55  296.12000  15.420013   
2 1948-01-03  18.470001  73.500015  102227.0  287.45  296.09998  13.070007   
3 1948-01-04  17.019989  62.500015  102082.0  287.05  296.07000   7.790008   
4 1948-01-05  13.940002  57.250015  102217.0  286.57  295.99000  19.050003   

       add6  
0  2.390015  
1 -1.259994  
2 -0.189987  
3 -0.259995  
4  3.150009

In [ ]:
len(x)*0.7

18132.1

In [ ]:
tr = x[:18132]
ts = x[18132:]

tr means training set, ts means testing set

In [ ]:
# for univariate
train_df = tr[['ds','y']]
test_df = ts[['ds']]
# for multivariate
TRAIN_df = tr[['ds','add1','add2','add3','add4','add5','add6','y']]
TEST_DF = ts[['ds','add1','add2','add3','add4','add5','add6']]

In [ ]:
# Univariate
Test_df = ts[['ds','y']]
# Multivariate
TEST_df = ts[['ds','add1','add2','add3','add4','add5','add6','y']]

In [ ]:
train_df.reset_index(inplace = True,drop = True)
test_df.reset_index(inplace = True, drop = True)
TRAIN_df.reset_index(inplace = True, drop = True)
TEST_DF.reset_index(inplace = True, drop = True)
Test_df.reset_index(inplace = True, drop = True)
TEST_df.reset_index(inplace = True, drop = True)

In [ ]:
Test_df.head()

index         ds          y
0  18132 1997-08-23  38.970000
1  18133 1997-08-24  43.960007
2  18134 1997-08-25  60.190002
3  18135 1997-08-26  74.459990
4  18136 1997-08-27  67.970000

#### I am done with preprocessing the data. Time forecast with fb prophet univariate and multivariate forecast

In [ ]:
!pip install pystan~=2.14
!pip install fbprophet 

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 67.3 MB 136 kB/s 
  Attempting uninstall: pystan
    Found existing installation: pystan 3.3.0
    Uninstalling pystan-3.3.0:
      Successfully uninstalled pystan-3.3.0
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 64 kB 1.8 MB/s 
  Created wheel for fbprophet: filename=fbprophet-0.7.1-py3-none-any.whl size=6637961 sha256=c4a5be56d990dd153cbb630c3e1cdb559149bc117f12596aba006198b8750c21
  Stored in directory: /root/.cache/pip/wheels/cd/a1/12/db63ff624de492fe6cccf676091a0860fdde2ffde4bc3280e2
Successfully built fbprophet
  Attempting uninstall: cmdstanpy
    Found existing installation: cmdstanpy 1.0.7
    Uninstalling cmdstanpy-1.0.7:
      Successfully uninstalled cmdstanpy-1.0.7
ERROR: pip's dependency resolver does not currently take into account all the 

In [ ]:
from fbprophet import Prophet

#### Univariate forecast 

In [ ]:
# Univariate forecast
model1 = Prophet()
model1.fit(train_df)

INFO:fbprophet:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


In [ ]:
forecast1 = model1.predict(test_df)

In [ ]:
forecast1 = forecast1[['ds', 'yhat']]

In [ ]:
final_pred1 = pd.concat((forecast1['yhat'],Test_df), axis = 1 )
final_pred1.head()

yhat  index         ds          y
0  44.184190  18132 1997-08-23  38.970000
1  43.833407  18133 1997-08-24  43.960007
2  43.398976  18134 1997-08-25  60.190002
3  43.020283  18135 1997-08-26  74.459990
4  42.610494  18136 1997-08-27  67.970000

In [ ]:
import matplotlib.pyplot as plt

#### Evaluating the univariate model

In [ ]:
Evaluation_metrics(final_pred1.yhat, final_pred1.y)

The Mean Absolute Error is  4.87
The Mean Squared Error is  41.73
The Root Mean Squared Error is  6.46
The Mean Absolute Percentage Error is  4.87


### Multivariate forecast

In [ ]:
model2 = Prophet()
model2.add_regressor('add1', standardize = False)
model2.add_regressor('add2', standardize = False)
model2.add_regressor('add3', standardize = False)
model2.add_regressor('add4', standardize = False)
model2.add_regressor('add5', standardize = False)
model2.add_regressor('add6', standardize = False)
model2.fit(TRAIN_df)

INFO:fbprophet:Disabling daily seasonality. Run prophet with daily_seasonality=True to override this.


In [ ]:
forecast2 = model2.predict(TEST_DF)

In [ ]:
forecast2 = forecast2[['ds', 'yhat']]

In [ ]:
final_pred2 = pd.concat((forecast2['yhat'],TEST_df), axis = 1 )
final_pred2.head()

yhat  index         ds       add1      add2    add3       add4  \
0  39.981772  18132 1997-08-23  38.250000  100370.0  301.87  300.64000   
1  41.456520  18133 1997-08-24  41.500000  100115.0  302.12  300.59998   
2  48.723079  18134 1997-08-25  62.250015   99892.0  301.40  300.59998   
3  57.272174  18135 1997-08-26  89.500015   99775.0  298.70  300.55000   
4  56.298637  18136 1997-08-27  88.750015  100115.0  299.20  300.40000   

       add5      add6          y  
0 -6.389984 -5.609985  38.970000  
1 -7.739990 -7.559998  43.960007  
2 -5.689987 -9.139984  60.190002  
3 -2.379990 -6.939987  74.459990  
4  1.770004 -2.309998  67.970000

In [ ]:
Evaluation_metrics(final_pred2.yhat, final_pred2.y)

The Mean Absolute Error is  2.98
The Mean Squared Error is  14.54
The Root Mean Squared Error is  3.81
The Mean Absolute Percentage Error is  2.98


#### Neural Prophet

In [ ]:
! pip install neuralprophet


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 98 kB 3.4 MB/s 
     |████████████████████████████████| 1.6 MB 29.7 MB/s 


In [ ]:
from neuralprophet import NeuralProphet

#### Univariate forecast

In [ ]:
train_df1 = pd.concat((train_df['y'],ds_df), axis = 1 )

In [ ]:
Test_df2 = pd.concat((Test_df1['y'],ds2_df), axis = 1 )

#### The reason for the above lines of codes is that for neural prophet to train the model to work, it needs to be arranged as shown below. So the codes were just my attempt at arranging them which turned out fine

In [ ]:
Test_df2.head(2)

y         ds
0  38.970000 1997-08-23
1  43.960007 1997-08-24

In [ ]:
y_df = train_df['y']

In [ ]:
ds_df = train_df['ds'] # This is just me isolating the time column so i can concatenate it again in the accepted arrangement

In [ ]:
model3 = NeuralProphet()
model3.fit(train_df1)

INFO - (NP.df_utils._infer_frequency) - Major frequency D corresponds to 99.994% of the data.
INFO:NP.df_utils:Major frequency D corresponds to 99.994% of the data.
INFO - (NP.df_utils._infer_frequency) - Dataframe freq automatically defined as D
INFO:NP.df_utils:Dataframe freq automatically defined as D
INFO - (NP.config.init_data_params) - Setting normalization to global as only one dataframe provided for training.
INFO:NP.config:Setting normalization to global as only one dataframe provided for training.
INFO - (NP.utils.set_auto_seasonalities) - Disabling daily seasonality. Run NeuralProphet with daily_seasonality=True to override this.
INFO:NP.utils:Disabling daily seasonality. Run NeuralProphet with daily_seasonality=True to override this.
INFO - (NP.config.set_auto_batch_epoch) - Auto-set batch_size to 64
INFO:NP.config:Auto-set batch_size to 64
INFO - (NP.config.set_auto_batch_epoch) - Auto-set epochs to 88
INFO:NP.config:Auto-set epochs to 88


  0%|          | 0/156 [00:00<?, ?it/s]

INFO - (NP.utils_torch.lr_range_test) - lr-range-test results: steep: 1.09E-01, min: 3.53E+00
INFO:NP.utils_torch:lr-range-test results: steep: 1.09E-01, min: 3.53E+00


  0%|          | 0/156 [00:00<?, ?it/s]

INFO - (NP.utils_torch.lr_range_test) - lr-range-test results: steep: 1.25E-01, min: 2.71E+00
INFO:NP.utils_torch:lr-range-test results: steep: 1.25E-01, min: 2.71E+00
INFO - (NP.forecaster._init_train_loader) - lr-range-test selected learning rate: 1.11E-01
INFO:NP.forecaster:lr-range-test selected learning rate: 1.11E-01
Epoch[88/88]: 100%|██████████| 88/88 [01:03<00:00,  1.38it/s, SmoothL1Loss=0.00663, MAE=4.5, RMSE=5.83, Loss=0.00492, RegLoss=0]


SmoothL1Loss        MAE        RMSE      Loss  RegLoss
0       1.421884  96.354740  109.285123  1.109386      0.0
1       0.629266  53.911236   62.077762  0.477470      0.0
2       0.062898  14.770999   17.358611  0.040290      0.0
3       0.022061   8.489463   10.608768  0.014138      0.0
4       0.011432   5.962000    7.637855  0.007620      0.0
..           ...        ...         ...       ...      ...
83      0.006713   4.534190    5.859323  0.004978      0.0
84      0.006680   4.522298    5.845281  0.004953      0.0
85      0.006663   4.514307    5.843954  0.004938      0.0
86      0.006647   4.511879    5.837812  0.004924      0.0
87      0.006634   4.501751    5.827151  0.004916      0.0

[88 rows x 5 columns]

In [ ]:
forecast3 = model3.predict(Test_df2)

INFO - (NP.df_utils._infer_frequency) - Major frequency D corresponds to 99.987% of the data.
INFO:NP.df_utils:Major frequency D corresponds to 99.987% of the data.
INFO - (NP.df_utils._infer_frequency) - Defined frequency is equal to major frequency - D
INFO:NP.df_utils:Defined frequency is equal to major frequency - D
INFO - (NP.df_utils._infer_frequency) - Major frequency D corresponds to 99.987% of the data.
INFO:NP.df_utils:Major frequency D corresponds to 99.987% of the data.
INFO - (NP.df_utils._infer_frequency) - Defined frequency is equal to major frequency - D
INFO:NP.df_utils:Defined frequency is equal to major frequency - D
INFO - (NP.df_utils.return_df_in_original_format) - Returning df with no ID column
INFO:NP.df_utils:Returning df with no ID column


In [ ]:
forecast3.head()

ds          y  residual1      yhat1      trend  season_yearly  \
0 1997-08-23  38.970000   5.232904  44.202904  24.461899      19.766748   
1 1997-08-24  43.960007  -0.147141  43.812866  24.461548      19.391178   
2 1997-08-25  60.190002 -16.733051  43.456951  24.461191      19.007242   
3 1997-08-26  74.459990 -31.343398  43.116592  24.460840      18.615332   
4 1997-08-27  67.970000 -25.242614  42.727386  24.460485      18.215836   

   season_weekly  
0      -0.025742  
1      -0.039861  
2      -0.011484  
3       0.040422  
4       0.051066

#### Evaluating metrics

In [ ]:
Evaluation_metrics(forecast3.yhat1, forecast3.y)

The Mean Absolute Error is  4.92
The Mean Squared Error is  42.56
The Root Mean Squared Error is  6.52
The Mean Absolute Percentage Error is  4.92


#### Multivariate forecast

In [ ]:
TRAIN_df1 = tr[['y','ds','add1','add2','add3','add4','add5','add6']]
TEST_DF1 = ts[['y','ds','add1','add2','add3','add4','add5','add6']]

In [ ]:
model4 = NeuralProphet()
model4.add_lagged_regressor('add1')
model4.add_lagged_regressor('add2')
model4.add_lagged_regressor('add3')
model4.add_lagged_regressor('add4')
model4.add_lagged_regressor('add5')
model4.add_lagged_regressor('add6')
model4.fit(TRAIN_df1)

INFO - (NP.forecaster.add_lagged_regressor) - n_lags = 'auto', but there is no lags for Autoregression. Number of lags for regressor is automatically set to 1
INFO:NP.forecaster:n_lags = 'auto', but there is no lags for Autoregression. Number of lags for regressor is automatically set to 1
INFO - (NP.forecaster.add_lagged_regressor) - n_lags = 'auto', but there is no lags for Autoregression. Number of lags for regressor is automatically set to 1
INFO:NP.forecaster:n_lags = 'auto', but there is no lags for Autoregression. Number of lags for regressor is automatically set to 1
INFO - (NP.forecaster.add_lagged_regressor) - n_lags = 'auto', but there is no lags for Autoregression. Number of lags for regressor is automatically set to 1
INFO:NP.forecaster:n_lags = 'auto', but there is no lags for Autoregression. Number of lags for regressor is automatically set to 1
INFO - (NP.forecaster.add_lagged_regressor) - n_lags = 'auto', but there is no lags for Autoregression. Number of lags for regr

  0%|          | 0/156 [00:00<?, ?it/s]

INFO - (NP.utils_torch.lr_range_test) - lr-range-test results: steep: 8.37E-02, min: 1.39E+00
INFO:NP.utils_torch:lr-range-test results: steep: 8.37E-02, min: 1.39E+00


  0%|          | 0/156 [00:00<?, ?it/s]

INFO - (NP.utils_torch.lr_range_test) - lr-range-test results: steep: 1.09E-01, min: 1.39E+00
INFO:NP.utils_torch:lr-range-test results: steep: 1.09E-01, min: 1.39E+00
INFO - (NP.forecaster._init_train_loader) - lr-range-test selected learning rate: 9.71E-02
INFO:NP.forecaster:lr-range-test selected learning rate: 9.71E-02
Epoch[88/88]: 100%|██████████| 88/88 [01:37<00:00,  1.10s/it, SmoothL1Loss=0.00444, MAE=3.7, RMSE=4.77, Loss=0.00331, RegLoss=0]


SmoothL1Loss        MAE       RMSE      Loss  RegLoss
0       0.471543  43.319477  53.746793  0.348187      0.0
1       0.144819  21.645737  27.028493  0.106161      0.0
2       0.045653  12.006529  15.257391  0.033681      0.0
3       0.025347   8.913752  11.364569  0.018955      0.0
4       0.013353   6.540692   8.253984  0.009977      0.0
..           ...        ...        ...       ...      ...
83      0.004515   3.732282   4.812570  0.003365      0.0
84      0.004485   3.717325   4.796670  0.003341      0.0
85      0.004463   3.711253   4.782940  0.003324      0.0
86      0.004454   3.705111   4.775721  0.003318      0.0
87      0.004439   3.698947   4.771843  0.003306      0.0

[88 rows x 5 columns]

In [ ]:
forecast4 = model4.predict(TEST_DF1)

INFO - (NP.df_utils._infer_frequency) - Major frequency D corresponds to 99.987% of the data.
INFO:NP.df_utils:Major frequency D corresponds to 99.987% of the data.
INFO - (NP.df_utils._infer_frequency) - Defined frequency is equal to major frequency - D
INFO:NP.df_utils:Defined frequency is equal to major frequency - D
INFO - (NP.df_utils._infer_frequency) - Major frequency D corresponds to 99.987% of the data.
INFO:NP.df_utils:Major frequency D corresponds to 99.987% of the data.
INFO - (NP.df_utils._infer_frequency) - Defined frequency is equal to major frequency - D
INFO:NP.df_utils:Defined frequency is equal to major frequency - D
INFO - (NP.df_utils.return_df_in_original_format) - Returning df with no ID column
INFO:NP.df_utils:Returning df with no ID column


In [ ]:
Evaluation_metrics(forecast4.yhat1, forecast4.y)

The Mean Absolute Error is  3.64
The Mean Squared Error is  22.71
The Root Mean Squared Error is  4.77
The Mean Absolute Percentage Error is  3.64


### Summary of results

In [ ]:
metrics = {'Model' : ['Univariate fbprophet', 'Multivariate fbprophet', 'Univariate neural prophet', 'Multivariate Neural prophet'], 'MAE' : [4.87,2.98,4.92,3.64], 'MSE': [41.73,14.54,42.56,22.71], 'RMSE':[6.46,3.81,6.52,4.77],'MAPE':[4.87,2.98,4.92,3.64]}

In [ ]:
Evaluation_metrics = pd.DataFrame(metrics)
display(Evaluation_metrics.head())

Model   MAE    MSE  RMSE  MAPE
0         Univariate fbprophet  4.87  41.73  6.46  4.87
1       Multivariate fbprophet  2.98  14.54  3.81  2.98
2    Univariate neural prophet  4.92  42.56  6.52  4.92
3  Multivariate Neural prophet  3.64  22.71  4.77  3.64

In [ ]:
Evaluation_metrics.columns

Index(['Model', 'MAE', 'MSE', 'RMSE', 'MAPE'], dtype='object')

In [ ]:
px.bar(Evaluation_metrics, x = 'Model', y = 'MAE',title = 'Mean Absolute Error Plot')

In [ ]:
px.bar(Evaluation_metrics, x = 'Model', y = 'MSE',title = 'Mean Squared Error Plot')

In [ ]:
px.bar(Evaluation_metrics, x = 'Model', y = 'RMSE',title = 'Root Mean Squared Error Plot')

In [ ]:
px.bar(Evaluation_metrics, x = 'Model', y = 'MAPE',title = 'Mean Absolute Percentage Error Plot')

#### Clearly from the plot you can see that the model that performed the best was multivariate fb prophet. From this project also, we can also infer that adding regressors or using more than one feature increases accuracy by a lot. Generally with this data, the fb prophet performed better than the neural prophet although i didn't do any parameter tuning. But I really don't think that would have changed much. 